node/coinstacks/common/api/src/registry.ts

```diff
@@ -0,0 +1,135 @@
+type BlockHandler<T = any, T2 = any> = (block: T) => Promise<{ txs: T2 }>
+/* eslint-disable-next-line @typescript-eslint/no-explicit-any */
+type TransactionHandler<T = any, T2 = any> = (tx: T) => Promise<{ addresses: Array<string>; tx: T2 }>
+
+export class Registry {
+  private clients: Record<string, Map<string, void>>
+  private addresses: Record<string, Map<string, ConnectionHandler>>
+
+  private formatAddressFunc: FormatAddressFunc = (address: string) => address.toLowerCase()
+  private handleBlock?: BlockHandler
+  private handleTransaction?: TransactionHandler
+
+  private logger = new Logger({ namespace: ['unchained', 'common', 'api', 'registry'], level: process.env.LOG_LEVEL })
+
+  constructor() {
+    this.clients = {}
+    this.addresses = {}
+  }
+
+  private toId(clientId: string, subscriptionId: string): string {
+    return `${clientId}:${subscriptionId}`
+  }
+
+  private fromId(id: string): { clientId: string; subscriptionId: string } {
+    const [clientId, subscriptionId] = id.split(':')
+    return { clientId, subscriptionId }
+  }
+
+  formatAddress(func: FormatAddressFunc): Registry {
+    this.formatAddressFunc = func
+    return this
+  }
+
+  transactionHandler<T, T2>(func: TransactionHandler<T, T2>): Registry {
+    this.handleTransaction = func
+    return this
+  }
+
+  blockHandler<T, T2>(func: BlockHandler<T, T2>): Registry {
+    this.handleBlock = func
+    return this
+  }
+
+  subscribe(clientId: string, subscriptionId: string, connection: ConnectionHandler, addresses: Array<string>) {
+    const id = this.toId(clientId, subscriptionId)
+
+    if (!this.clients[id]) this.clients[id] = new Map<string, void>()
+
+    addresses.forEach((address) => {
+      address = this.formatAddressFunc(address)
+
+      if (!this.addresses[address]) this.addresses[address] = new Map<string, ConnectionHandler>()
+
+      this.clients[id].set(address)
+      this.addresses[address].set(id, connection)
+    })
+  }
+
+  unsubscribe(clientId: string, subscriptionId: string, addresses: Array<string>) {
+    const id = this.toId(clientId, subscriptionId)
+
+    if (!this.clients[id]) return
+
+    const unregister = (id: string, address: string) => {
+      address = this.formatAddressFunc(address)
+
+      // unregister address from client
+      this.clients[id].delete(address)
+
+      // delete client from registery if no addresses are registered anymore
+      if (!this.clients[id].size) delete this.clients[id]
+
+      if (this.addresses[address]) {
+        // unregister client from address
+        this.addresses[address].delete(id)
+
+        // delete address from registery if no clients are registered anymore
+        if (!this.addresses[address].size) delete this.addresses[address]
+      }
+    }
+
+    if (!addresses.length) {
+      for (const address of this.clients[id].keys()) {
+        unregister(id, address)
+      }
+
+      delete this.clients[id]
```

Comment:
Looks like this is occuring in the unregister function from the size check. redundant